jslib/locomote.js

```diff
@@ -15,6 +15,66 @@
 (this, function() {
   'use strict';
 
+  function SyncMonitor(syncResolution) {
+    this.syncResolution = syncResolution || 1000;
+    this.players = {};
+  }
+
+  SyncMonitor.prototype = {
+    registerPlayer: function (player, initialTime) {
+      this.players[player.tag] = {
+        player: player,
+        initialTime: initialTime || 0,
+        buffer: [],
+        lastPlayed: -1,
+        startedAt: 0
+      };
+      player.config({ step: true });
+      player.on('frame:ready', this.addFrame.bind(this, player.tag));
+    },
+
+    unregisterPlayer: function (player) {
+      this.players[player.tag].player.off('frame:ready');
+      delete this.players[player.tag];
+    },
+
+    playFrameFor: function (id) {
+      var time = this.players[id].buffer.shift();
+      if (time || time === 0) {
+        this.players[id].lastPlayed = time;
+        this.players[id].player.playNextFrame();
+      }
+      return time;
+    },
+
+    addFrame: function (id, timestamp) {
+      this.players[id].buffer.push(timestamp);
+      setTimeout(this.tryPlayFrame.bind(this), 0);
+    },
+
+    tryPlayFrame: function () {
+      var ids = Object.keys(this.players);
+      var nextPlayer = ids.reduce(function (id, key) {
+        var min = this.players[id].buffer[0];
+        var current = this.players[key].buffer[0];
+        if (min === undefined || current < min) {
+          return key;
+        } else {
+          return id;
+        }
+      }.bind(this), ids[0]);
+
+      var mayPlay = ids.every(function (id) {
+        return id === nextPlayer || this.players[nextPlayer].buffer[0] - this.players[id].lastPlayed < this.syncResolution;
```

Comment:
this is really reliant on operator presedence which might not be obvious how it operates. Parenthesis would help understanding this statement